### PR TITLE
Prometheus: Configuration page overhaul

### LINKED ISSUE #65877: Prom Config Overhaul Epic
This issue is to organize the efforts around the Prometheus Configuration Page Overhaul.

[The work and meeting notes are here](https://docs.google.com/document/d/1Pw0l30ijhBAnXqm5IFPU8jPEtiL5ZtW29OQhE-dU5l8/edit#heading=h.3cdq44kqtkii)

The list of work includes Stages 1 & 2
[The discussion of this breakdown is here](https://docs.google.com/document/d/1WqEbJvInT_94xA4M-pATMYypnjGIs3dbwNbEvsqGuF8/edit#heading=h.rldiir2jr775), April 14, 2023

Stage 1 https://github.com/grafana/grafana/pull/66198

- [x] Implement health check, improve error handling for the test connection function
- [x] Organize additional settings
  - [x] Meet with Galen and Ismail
  - [x] Consider refactoring complex components 
- [x] Improve tooltips for each element and add links to docs
- [x] Add custom labels for the connection URL 

Stage 2
- [x] build behind feature flag to test Azure and SigV4 auth
- [x] [Description component](https://docs.google.com/document/d/11XMaYHSMSra8AoN2-yVW-5QPbjEXgkAFbTIF6N84cGE/edit#heading=h.o1zp7h8irwci)
- [x] Connection URL
- [x] Auth 
- [x] Use new auth component 
- [x] Improve error handling
  - [x] Return more verbose errors
  - [ ] Return success with Prom type. The prom type image used to be shown in the save and test modal message
- [ ] Use Editable header component
- [x] Toggle open/close/hide settings that are not the bare minimum

### LINKED ISSUE #65970: /api/datasources/uid/:uid/health api doesn't work for prometheus datasources
Hi Grafana Team,

Can someone clarify why the below datasource health api results in "message: Not found" ? for prometheus datasource. The same works fine for influxdb datasource though. When I looked at the network calls from UI for influxdb and prometheus datasource connectivity test , I don't see health api invoked for prometheus datasource unlike influxdb. But both were successful from UI. from API call's only influxdb health api can be verified.

The api seems to be correct as per the reference link : https://grafana.com/docs/grafana/latest/developers/http_api/data_source/#check-data-source-health

Screen Shot 2023-04-04 at 1 23 08 PM

I could also confirm removing "health" at the end does return the datasource details.:

Screen Shot 2023-04-04 at 1 26 30 PM

### LINKED ISSUE #66829: When saving a Prometheus datasource the /api/v1/status/buildinfo req doesn't contain the Authorization header
## What happened:
On the Prometheus datasource page, when I click on `Save & test` then a new request is initiated to Prometheus' `/api/v1/status/buildinfo` endpoint. Unfortunately, it doesn't go through the `AzureMiddleware`thus it doesn't contain the mandatory `Authorization` header. It can cause a slight issue (it doesn't break the functionality) when there is a auth proxy in front of Prometheus.

## What you expected to happen:
The outgoing request to the `/api/v1/status/buildinfo` contains the `Authorization` header.

## How to reproduce it (as minimally and precisely as possible):
Go to the Prometheus datasource page, setup the datasource, then click on `Save & test` then a new request is initiated to Prometheus' `/api/v1/status/buildinfo` endpoint. 

You can use the `auth/prometheus_oauth2_proxy_azure` devenv (make sure to remove the `skip_auth_routes` from the oauth2-proxy.cfg)
## Environment:
- Grafana version: 10.0.0-pre
- Data source type & version: Prometheus 2.43.0
- OS Grafana is installed on: Mac (development system)
- User OS & Browser: Mac, Chrome
- Grafana plugins: Prometheus
- Others: "Azure Authentication" is configured on the datasource page

### PR DESCRIPTION
**Pull request by bohandley**
**What is this?**
[Config Overhaul Stage 1](https://docs.google.com/document/d/1WqEbJvInT_94xA4M-pATMYypnjGIs3dbwNbEvsqGuF8/edit#) This is a first stage of the [Prometheus Config Overhaul Epic](https://github.com/grafana/grafana/issues/65877) as driven by G10 Getting Started and the Config Overhaul Working Group. These changes were described and shared to me by @josiahg, @catherineymgui and @gtk-grafana 

- [x] See discussion below with @torkelo and @catherineymgui, we are adding a prom label and interactive tooltip to the URL input in DatasourceHttpSettting.ts instead of creating a new custom bare minimum section. The bare minimum section will be added in stage 2 once the auth component has been refactored and the styles are consistent. For now, we are incrementally moving toward the config overhaul working group goal of making the UI more clear with Prom label/adding doc tooltip for setting up a Prometheus datasource.
Screenshot 2023-04-26 at 3 28 41 PM

  - auth will be a new component added in stage 2 and design consistency will be updated
  - Connection section will be created in stage 2 with the auth and design with consistent padding will be completed.
  
- [x] organizing other config elements into categories to not clutter the page (maybe hide additional elements)
Screenshot 2023-04-17 at 3 04 33 PM

- [x] Save & Test: adding test backend datasource health check to improve error handling(url and auth)
  - Examples
  - [x] implement `/-/healthy` api for prometheus types that support this api
Screenshot 2023-04-16 at 6 58 17 PM

  - [x] Url error with timeout
  Screenshot 2023-04-17 at 2 41 27 PM
  Screenshot 2023-04-17 at 2 41 47 PM
  Screenshot 2023-04-17 at 2 55 22 PM

  - [x] Simple url error 
  Screenshot 2023-04-17 at 2 42 30 PM
  Screenshot 2023-04-17 at 2 50 38 PM

  - [x] fallback to a query for instances that do not implement the `/-/healthy` api
Screenshot 2023-04-16 at 6 59 21 PM

  - [x] incorrect url for instance that does not implement the `/-/healthy` api
Screenshot 2023-04-16 at 8 53 08 PM

  - [x] incorrect authentication for instance that does not implement the `/-/healthy` api
Screenshot 2023-04-16 at 9 02 34 PM

- [x] adding links to  documentation
- [x] updating tooltips for config elements
- [x] removing legacy form inputs

This work aligns with the Docs Overhaul I am working on with Larissa Wandzura on the docs team. We are updating the Prometheus Configuration Docs for the overhaul. https://grafana.com/docs/grafana/latest/datasources/prometheus/#configure-the-data-source

The UI suggestions, feature updates and organization plans are listed in this doc which has assistance from @itsmylife and others mentioned above. https://docs.google.com/document/d/1Pw0l30ijhBAnXqm5IFPU8jPEtiL5ZtW29OQhE-dU5l8/edit#

**Why do we need this feature?**
G10 Getting Started: Prometheus research and the Config working group uncovered issues with datasource configuration pages in Grafana. People can get stuck setting up datasources. 90% of people click save and test but 4% are successful on start. The work suggested will help improve people's success in setting up datasources.

**Who is this feature for?**
Grafana users who are setting up a Prometheus datasource.

**Which issue(s) does this PR fix?**:
Fixes https://github.com/grafana/grafana/issues/65877
Fixes https://github.com/grafana/grafana/issues/65970
Fixes https://github.com/grafana/grafana/issues/66829

**Special notes for your reviewer:**
This is a first stage. The tooltips may need more updating, the links to docs need updated by the work done with Larissa, the auth component and configurable header component are being worked on by @grafana/enterprise-datasources, and error handling from testing the datasource needs updated (another PR for creating a backend health check which prometheus does not have).

The previous config page is here:
Screenshot 2023-04-08 at 2 41 00 PM

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by catherineymgui — DISMISSED** [review]
@josiahg and I took a look and had comments

UI:

- [x] 32px between the section headers
- [x] Section descriptions - use secondary grey colour
- [x] Subheaders under additional settings - use h6
- [x] Move the toggle in outside of the field in Manage alerts via Alerting UI and Disable metrics lookup
  - [X] *Brendan - most are done but need to figure out a nice way to do Alerting
- [x] It seems most of the form fields are using a legacy input field (which is why certain fields have different tooltip colours). Could you double check this
  - [X] *Brendan - most are done in PromSettings.ts and DatasourceHttpSettingsOverhaul.ts but there might still be some and I need to poke around some more. 

  Copy:
  
- [X] “See the documentation” link opens in the same tab instead of new tab, rips you out of Grafana. Also there’s an extra period
- [x] See documentation section — use secondary grey
- [ ] See documentation section — Note: We are adding an additional component (https://github.com/grafana/plugin-ui/pull/64) — this is in plugin-ui now, but will be added to core. I don’t know if you want to remove this for now
  - [ ] *Brendan - holding off on this one for now until it's in core grafana
- [x] Potential renaming of groups: Prometheus Behavioural —> Interval Behaviour, UI —> Query Editor, Advanced —> Other
- [x] While we are not perfectly consistent, prefer “data source” to “datasource” throughout the page
- [x] URL -> “Prometheus Server URL”
- [x] Tooltip: “This URL must be accessible from the Grafana server”
- [x] Add tooltip for “manage alerts via Alerting UI” — Matias made one here https://github.com/grafana/grafana/pull/66216#event-8965241606
- [x] Prometheus Behavioural - tooltips don’t help me determine this (where do i look in Prom to find this info?) they also don’t let me know if there are any consequences if I do it wrong - if I have my Prom set for 10s and i set this config to 60s what happens?
- [x] “Default editor” tooltip needs updated to - “Set default editor option for all users of this data source. If no option was selected, the default editor will be the "builder".”
- [x] Can we just set the default to Builder in the config UI and delete the second sentence as well?
- [x] Prometheus type - no information about implications if we do this wrong
- [x] Custom query parameters - what is the reason you would want to do this? (Contrast with HTTP method tooltip which is quite good and explains implications)
- [x] Exemplars - Internal Link - no tooltip, what does this setting do?
- [x] Would be good to ask Larissa to take a look to ensure consistency in content, and tooltips

**Comment by bohandley** [comment]
Need to work on the following inputs that may be shared across other datasources. I need to either recreate them with the right design(InlineField over FormField and Switch over InlineSwitch)
Screenshot 2023-04-12 at 8 56 08 PM
Screenshot 2023-04-12 at 8 56 16 PM

Need to recreate validation events for `Scrape Interval`, `Query Timeout`, and `Query overlap window` for incremental querying because only legacy Input component accepts validation events. There should be another way to validate the Input component value from grafana-data.

**Comment by sasklacz** [comment]
@bohandley as for the error handling, for now please hold this off as we plan to rewrite one of the plugins later this month to show how this could be done taking into account all of our previous ideas and conclusions. So realistically we can start shipping those plugins sometime in May.

**Comment by itsmylife** [comment]
@bohandley Following image is not related to your work but we don't really show the details. Maybe in this PR, you wanna address this?
image

**Comment by itsmylife** [comment]
I am using a wide monitor and the config page looks like this on my screen. In small monitors, they go line by line but this way I cannot determine which toggle is for which setting. It is kinda a bad experience for me. 
image

**Comment by itsmylife** [comment]
Also labels and input fields are not aligning. I thought this can be also addressed in this PR. If that's not the intention we can open another issue to address this. 
image

**Comment by itsmylife** [comment]
Also another aligning issue picture 
image

**Comment by itsmylife** [comment]
So maybe @catherineymgui can be the right person for this kind of question. 
Referring the thread https://raintank-corp.slack.com/archives/C0F5GDR4P/p1681224647575209

**Comment by cletter7** [comment]
Yes, I am currently working on the Auth component designed by @catherineymgui that should replace the `DataSourceHttpSettings`

**Comment by bohandley** [comment]
Thank you @catherineymgui @sasklacz @cletter7!

@yesoreyeram and @itsmylife Catherine, Kuba, Andriy and myself are meeting today to sync up about this work. Much of this PR will be pulled back/removed, so @josiahg there will be major design differences. I will update the things I can based on the meeting. Thank you everyone!

**Comment by bohandley** [comment]
I removed the duplicate efforts @yesoreyeram @cletter7 , I made the elements that we have control over have consistent label width @itsmylife, I added interactive tooltips to click on links in tooltips @josiahg. This will be stage one according to our meeting earlier today. Stage 2 will be where I use the auth component that is being developed in plugins-ui and wrap that change behind a feature toggle. https://docs.google.com/document/d/1WqEbJvInT_94xA4M-pATMYypnjGIs3dbwNbEvsqGuF8/edit?usp=sharing

@josiahg after our meeting I was able to get the save and test to return a false positive so I will be working on the health check for the backend. This will be in another PR though because I will need @grafana/observability-metrics to weigh in on the depth of what the health check will test.

Thanks all, have a great weekend!

**Comment by bohandley** [comment]
Adding health check to this PR so that we can get it in before G10 cutoff on May 1

**Comment by bohandley** [comment]
The simple URL check is not working currently, need to fix this. There are two responses for this, either a timeout or a false positive where a response is sent but the response is a redirect or some html from the network.

**Comment by catherineymgui** [comment]
@bohandley 

Let me know if these error / success messages are accurate:

Success

- Successfully saved the configuration and queried the Prometheus API.

Timeout
- Error querying resource: Get “[http://this./-/healthy”:](http://this./-/healthy%E2%80%9D:) dial tcp 143.244.220.150.80: io timeout  - Your configuration has been saved but there was a timeout. Try saving again.

Simple URL
- Your configuration has been saved but there is an error. Double check the URL formatting.

Cannot connect to the URL
- Your configuration has been saved but there is an error. Double check the URL formatting.

  Incorrect URL / 404
- Your configuration has been saved but there is an error. Double check the URL formatting .

Incorrect authentication / 401
- 401 Unauthorized - Your configuration has been saved but there is an error. Double check the authentication.

File: public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx:19]: will multiple duration types work? i.e `1h 5m 2s`?

File: public/app/plugins/datasource/prometheus/configuration/Connection.tsx
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/configuration/Connection.tsx:93]: we've got some inline styles, and some class names, can we make this a bit more consistent?

File: public/app/plugins/datasource/prometheus/configuration/Connection.tsx
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/configuration/Connection.tsx:52]: Where did we get this from? Is this currently in use in grafana anywhere? Do we have any additional documentation as to what urls this accepts?

Maybe unit tests for this if it's not already tested

**Review by gtk-grafana — DISMISSED** [review]
The FE looks good, a few small nits.
Including we've got a bunch of selects that aren't getting the disabled style for provisioned sources, and users can interact and (appear to) change their values.
image

File: public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx:19]: Good catch! No it does not, this was a custom prom settings duration validation
```
export const promSettingsValidationEvents = {
  [EventsWithValidation.onBlur]: [
    regexValidation(
      /^$|^\d+(ms|[Mwdhmsy])$/,
      'Value is not valid, you can use number with time unit specifier: y, M, w, d, h, m, s'
    ),
  ],
};
```

This is different from the `isValidDuration` duration check that you imported for incremental querying
```
/**
 * isValidDuration returns true if the given string can be parsed into a valid Duration object, false otherwise
 *
 * @param durationString - string representation of a duration
 *
 * @public
 */
export function isValidDuration(durationString: string): boolean {
  for (const value of durationString.trim().split(' ')) {
    const match = value.match(/(\d+)(.+)/);
    if (match === null || match.length !== 3) {
      return false;
    }

    const key = Object.entries(durationMap).find(([_, abbreviations]) => abbreviations?.includes(match[2]))?.[0];
    if (!key) {
      return false;
    }
  }

  return true;
}
```

For config items 'Query timeout' and 'Scrape interval' we never supported multiple durations. For incremental querying, do we support multiple durations?

File: public/app/plugins/datasource/prometheus/configuration/Connection.tsx
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/configuration/Connection.tsx:93]: Can do!

File: public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx:19]: Yeah but just because that function does, if this is the desired functionality then we should be good

File: public/app/plugins/datasource/prometheus/configuration/Connection.tsx
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/configuration/Connection.tsx:52]: This connection component logic was copied from https://github.com/grafana/grafana/blob/39a3d85514553e87bd17e516bcf97373e14ad9f6/packages/grafana-ui/src/components/DataSourceSettings/DataSourceHttpSettings.tsx#L122-L124 which implements this regex... but it does not test it 😆 I will write some unit tests for it.

**Comment by gtk-grafana** [comment]
Still lots of title case copy, can we switch to sentence casing?
image

**Comment by bohandley** [comment]
@gtk-grafana @catherineymgui @torkelo 

> Still lots of title case copy, can we switch to sentence casing? image

Question for sentence casing in Config headers, the example I was given for headers used title casing and not sentence casing. Which should I use, title or sentence?

Screenshot 2023-04-25 at 11 15 54 AM

**Comment by catherineymgui** [comment]
@bohandley 

Sorry, I totally missed this one. Please use sentence casing in all headers and labels as seen in the Figma designs!

**Comment by bohandley** [comment]
From @torkelo 
> Seems like very very inconsistent amount and form of spacing. Why the huge margins and separator line just for the URL property?

@catherineymgui @josiahg The config overhaul group requested that we have a "bare minimum" section that is separate from all other elements and the requested form of separation was the separator line with padding space of 32px. The bare minimum for Prometheus is a connection URL.  Other datasources require auth as part of the bare minimum and you can see that in the Oracle config. Ultimately I want to use padding instead of margin css in all components but as I am waiting on components being developed by enterprise squad folk I am using margins which allow for the desired space between the `<hr /> tags that are used for line separators (You can't add padding to those).

Screenshot 2023-04-12 at 15 51 42

**Comment by torkelo** [comment]
Name is also required, just looked a bit very strange (especially the line and border under Name). And 32px is more spacing between elements that I have seen we use for anything.

File: public/app/plugins/datasource/prometheus/configuration/Connection.tsx
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/configuration/Connection.tsx]: tests added! https://github.com/grafana/grafana/pull/66198/commits/7ef949ccefdbaa1bad51b78a717b009baaba6562

File: public/app/plugins/datasource/prometheus/configuration/Connection.tsx
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/configuration/Connection.tsx]: updated! https://github.com/grafana/grafana/pull/66198/commits/df31e475db0cabd7998f674125a18fc8e51e249e

File: public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx
Comment (bohandley) [reply public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx:21]: updated validation for incremental querying! https://github.com/grafana/grafana/pull/66198/commits/b9a0f01e3c9b25a1ac39c33b53a1d5d9f419e350

**Comment by catherineymgui** [comment]
@torkelo We're only using the 32px spacing between sections (not each field) for enough white space between them. It is a bit tight without the spacing. As Brendan said, there are some components that have to be updated in the 2nd round of redesigns so there's more consistency with the connection and HTTP settings parts.

Fyi - the name field is getting pulled out into an editable header component.

**Comment by bohandley** [comment]
@catherineymgui 

I had to change the implementation a little and simplify the health check as I found out that that Mimir does not implement the `healthy` api. Considering this and considering 'cloud first', I am moving forward with simplifying the health check and doing what the folks on the Mimir team suggested, simply performing a query and returning the errors.

I suggest any particular information we express by the error messages themselves and we add a generic message. Due to the responses returned and how Prometheus backend returns errors for queries, the way that I could disambiguate the errors and add special messages for each error would be brittle. If there was one perfect route to handle all types and versions for Prometheus, this would work, but sadly that is not the case.

I strongly suggest using this generic error message.
"Your configuration has been saved but there is an error."

This would look like the following:

1. `401 Unauthorized - Your configuration has been saved but there is an error.`
2. `404 Not Found - Your configuration has been saved but there is an error.`
3. `Error querying resource: Get “[http://gobbledygook”:](http://this./-/healthy%E2%80%9D:) dial tcp 143.244.220.150.80: io timeout - Your configuration has been saved but there is an error.`

And for success, this one is perfect 🎉 

`Successfully saved the configuration and queried the Prometheus API.`

**Comment by torkelo** [comment]
@catherineymgui Just the spacing looks so different from any other large form we have, and on a small laptop seems very excessive 
Screenshot 2023-04-26 at 07 16 34

Compared to dashboard settings:

Screenshot 2023-04-26 at 07 17 40

Alert contacts 
Screenshot 2023-04-26 at 07 17 05

**Comment by bohandley** [comment]
@torkelo and @catherineymgui 

Here are three possible ways we can go with the Connection URL styling. Because this is stage 1 work I am inclined to do the first choice because we will return to the styling in stage 2 with better components and it required the least code changes. But, I am open to doing the other two options which make incremental movements towards the UI suggestions based on the research of the config overhaul working group. WDYT?

1. Use the previous DatasourceHttpSettings URL input
  - Pros: Less excessive spacing on small screens
  - Cons: Missing UI parts that add clarity to the Prometheus URL set up (docs link, Prom custom field label), uses legacy form component
Screenshot 2023-04-26 at 10 26 00 AM

2. Remove all custom space styling (excessive space) around Connection section
  - Pros: could look better on smaller screens, has all Prom clarity details like field label, required star, tooltip with link
  - Cons: the design looks a little off for this style, still looks like we need a little and I would need some UI direction in what space to add if we need it.)
Screenshot 2023-04-26 at 10 38 15 AM

3. Pass the Connection label input and link to docs to set on the tooltip into the DatasourceHttpSettings
  - Pros: adds clarity to setting up datasource, custom prom label and link to docs in the tooltip (interactive)
  - Cons: Can't show the required star on the label because DatasourceHttpSettings is using the legacy form field label + input.
Screenshot 2023-04-26 at 11 45 04 AM

**Comment by bohandley** [comment]
After speaking with @catherineymgui, we compromised for the Connection styling @torkelo pointed out with this choice, choice three from the above comment. Please let me know if there are concerns.

Screenshot 2023-04-26 at 3 28 41 PM

**Comment by torkelo** [comment]
@bohandley look ok 👍

File: pkg/tsdb/prometheus/healthcheck.go
Comment (kylebrandt) [inline pkg/tsdb/prometheus/healthcheck.go:67]: Since this part doesn't do any saving, and just health check, I think it should just be an error message regarding the health check.

**Review by gtk-grafana — APPROVED** [review]
Looks good, thanks for all your hard work on this Brendan!

File: pkg/tsdb/prometheus/healthcheck.go
Comment (bohandley) [reply pkg/tsdb/prometheus/healthcheck.go:19]: Adding this for work here https://github.com/grafana/observability-metrics-squad/issues/116. Thank you Kyle!

**Review by torkelo — COMMENTED** [review]
Dam this was merged before I review & approve :(

will open a Fix PR

**Comment by bohandley** [comment]
My apologies @torkelo, I thought the last reviews were the UI ones. I was incorrect and merged this too quickly. I am sorry.

Thank you for opening the fix for the DatasourceHttpSettings, I am reviewing that now.

File: public/app/plugins/datasource/prometheus/datasource.tsx
Comment (bohandley) [inline public/app/plugins/datasource/prometheus/datasource.tsx:1054]: @sasklacz We removed it in this PR

**Comment by torkelo** [comment]
@bohandley No worries, should not have to feel you need to wait for approval from me. Was it my mistake on being late with a review (or I should have marked the first review as request changes)

File: pkg/tsdb/prometheus/querydata/request.go
Comment (bohandley) [inline pkg/tsdb/prometheus/querydata/request.go:182]: @AndrejKiri here PR and the spot where we return errors for health check

**Pull request merged**
